### PR TITLE
Support experiment tracking with W&B

### LINKED ISSUE #212: [feature request] Support experiment tracking via W&B
This PR adds experiment tracking via W&B. Users can run commands like

```python
python train.py --algo ppo --env CartPole-v1 --track --wandb-project-name sb3
```

which would result in the following run:

https://wandb.ai/costa-huang/sb3/runs/25jaqi4d?workspace=user-costa-huang

**Additional context**

Recently, I have been thinking about tracking experiments for other RL libraries as well so that metrics can be easily compared (see https://github.com/vwxyzjn/cleanrl/issues/123); hence the reason for this PR. I am happy to keep this as a separate fork, but just wonder if you would like to have the same functionality in the main repo.

## Related feature request

One related feature request: can we also log the raw (un-averaged) episodic return instead of just the `rollout/ep_rew_mean`?

### PR DESCRIPTION
**Pull request by vwxyzjn**
## Description
closes #212

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -114,6 +116,16 @@
         help="Overwrite hyperparameter (e.g. learning_rate:0.01 train_freq:10)",
     )
     parser.add_argument("-uuid", "--uuid", action="store_true", default=False, help="Ensure that the run has a unique ID")
+    parser.add_argument(
+        "--track",
+        type=lambda x: bool(strtobool(x)),
```
Comment (araffin) [inline train.py:121]: please use `action="store_true", default=False,` as for the `--no-render` of the enjoy script (unless there is something I'm missing)

File: train.py [hunk 2]
```diff
@@ -153,6 +165,22 @@
     print("=" * 10, env_id, "=" * 10)
     print(f"Seed: {args.seed}")
 
+    if args.track:
+        import wandb
```
Comment (araffin) [inline train.py:169]: Please do a try/except with an informative message ;)

File: train.py [hunk 3]
```diff
@@ -153,6 +165,22 @@
...
+
+        print(args)
```
Comment (araffin) [inline train.py:171]:
```suggestion
```

debug?

**Review by araffin — COMMENTED** [review]
Thanks for the PR =)
In fact, that was something I wanted to add at some point. Could you also document the use in the readme and update the changelog?

File: train.py (hunk 2 above)
Comment (vwxyzjn) [reply train.py:169]: Are you saying just to print out "tracking this experiment with Weights and Biases"? What exception are you anticipating with try/except.

**Comment by vwxyzjn** [comment]
Also a quick question, I would like to also log raw episodic return instead of the average one `rollout/ep_rew_mean`. What do you think is the best way to do it?

File: train.py (hunk 2 above)
Comment (araffin) [reply train.py:169]: `except ImportError` "if you want to use W&B tracking, please install ...

File: train.py (hunk 1 above)
Comment (araffin) [reply train.py:121]: I would prefer being consistent with the rest of the project, so `--track` if you want to track, otherwise nothing (just like `-optimize`).

**Comment by vwxyzjn** [comment]
> in both cases, you need a callback and you need to be careful when using multiple envs.

This is interesting. I guess for Mutil-agent envs, this can be quite tricky. how do you currently log it?

**Comment by araffin** [comment]
> I guess for Mutil-agent envs, this can be quite tricky. how do you currently log it?

I tend to use an `EvalCallback` anyway, because the training performance is different from the true performance.
Otherwise, we just loop through the `infos` list (https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/base_class.py#L459) and update the buffer.
For the monitor files, they are handled by each separately and contains the timestamps which allow to order them afterward.

File: README.md [hunk 4]
```diff
@@ -196,6 +196,22 @@ Note that the default hyperparameters used in the zoo when tuning are not always
 
 When working with continuous actions, we recommend to enable [gSDE](https://arxiv.org/abs/2005.05719) by uncommenting lines in [utils/hyperparams_opt.py](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/utils/hyperparams_opt.py).
 
+
+## Experiment tracking
+
+We support tracking experiment data such as learning curves and hyperparameters via [Weights and Biases](https://wandb.ai).
```
Comment (araffin) [inline README.md:202]: maybe put a link to SB3 integration doc?

File: train.py [hunk 5]
```diff
@@ -2,7 +2,10 @@
 import difflib
 import importlib
 import os
+import time
 import uuid
+from ast import Raise
```
Comment (araffin) [inline train.py:7]: why is that needed? ast is for analysing code, no?

File: train.py [hunk 6]
```diff
@@ -2,7 +2,10 @@
...
+from distutils.util import strtobool
```
Comment (araffin) [inline train.py:8]: unused import

File: train.py [hunk 7]
```diff
@@ -153,6 +164,24 @@
     print("=" * 10, env_id, "=" * 10)
     print(f"Seed: {args.seed}")
 
+    if args.track:
+        try:
+            import wandb
+        except ImportError:
+            raise ImportError("if you want to use W&B tracking, please install W&B via `pip install wandb`")
```
Comment (araffin) [inline train.py:171]:
```suggestion
            raise ImportError("if you want to use Weights&Biases tracking, please install W&B via `pip install wandb`")
```

File: README.md (hunk 4 above)
Comment (vwxyzjn) [reply README.md:202]: The thing is we are not really using the SB3 WANDB integration. The reason is that `args` variable in `train.py` contains the hyper-parameter info and we kind of no longer need SB3 WANDB integration.

File: train.py (hunk 7 above)
Comment (vwxyzjn) [reply train.py]: This is re-written.

File: train.py (hunk 5 above)
Comment (vwxyzjn) [reply train.py]: So weird. I think VScode auto imported it...

File: README.md (hunk 4 above)
Comment (araffin) [reply README.md:202]: ok

File: train.py [hunk 8]
```diff
@@ -187,12 +216,21 @@
         no_optim_plots=args.no_optim_plots,
     )
 
+    from utils.callbacks import RawStatisticsCallback
```
Comment (araffin) [inline train.py:219]: for another PR, no?

File: README.md (hunk 4 above)
Comment (vwxyzjn) [reply README.md:202]: > and https://wandb.ai/openrlbenchmark/sb3/runs/1b65ldmh is a 404 for me :/

This is fixed

File: train.py (hunk 8 above)
Comment (vwxyzjn) [reply train.py]: Removed.

File: README.md (hunk 4 above)
Comment (vwxyzjn) [reply README.md:202]: > a comment is missing then ;)

Added a comment.

File: train.py [hunk 9]
```diff
@@ -114,6 +115,14 @@
...
     )
     parser.add_argument("-uuid", "--uuid", action="store_true", default=False, help="Ensure that the run has a unique ID")
+    parser.add_argument(
+        "--track",
+        action="store_true",
+        default=False,
+        help="if toggled, this experiment will be tracked with Weights and Biases",
+    )
+    parser.add_argument("--wandb-project-name", type=str, default="sb3", help="the wandb's project name")
+    parser.add_argument("--wandb-entity", type=str, default=None, help="the entity (team) of wandb's project")
```
Comment (araffin) [inline train.py:125]: it will automatically use the username from the api key in case it is None?

File: train.py [hunk 10]
```diff
@@ -153,6 +162,26 @@
...
+
+        run_name = f"{args.env}__{args.algo}__{args.seed}__{int(time.time())}"
+        run = wandb.init(
+            name=run_name,
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            config=vars(args),
+            sync_tensorboard=True,  # auto-upload sb3's tensorboard metrics
+            monitor_gym=True,  # auto-upload the videos of agents playing the game
+            save_code=True,  # optional
```
Comment (araffin) [inline train.py:181]: not sure if needed (and doesn't seem to work anyway)

File: train.py (hunk 9 above)
Comment (vwxyzjn) [reply train.py:125]: Yes.

**Review by araffin — APPROVED** [review]
minor comments, otherwise LGTM =)
